smart-alluxio/src/main/java/org/smartdata/alluxio/metric/fetcher/AlluxioNamespaceFetcher.java

```diff
@@ -0,0 +1,298 @@
+      this.currentBatch = new FileStatusInternalBatch(DEFAULT_BATCH_SIZE);
+      this.fs = fs;
+      this.deque.add(ROOT);
+    }
+
+    @Override
+    public void run() {
+      if (LOG.isDebugEnabled()) {
+        long curr = System.currentTimeMillis();
+        if (curr - lastUpdateTime >= 2000) {
+          LOG.debug(String.format(
+              "%d sec, numDirectories = %d, numFiles = %d, batchsInqueue = %d",
+              (curr - startTime) / 1000,
+              numDirectoriesFetched, numFilesFetched, batches.size()));
+          lastUpdateTime = curr;
+        }
+      }
+      String parent = deque.pollFirst();
+      if (parent == null) { // BFS finished
+        if (currentBatch.actualSize() > 0) {
+          try {
+            this.batches.put(currentBatch);
+          } catch (InterruptedException e) {
+            LOG.error("Current batch actual size = "
+                + currentBatch.actualSize(), e);
+          }
+          this.currentBatch = new FileStatusInternalBatch(DEFAULT_BATCH_SIZE);
+        }
+
+        if (this.batches.isEmpty()) {
+          this.isFinished = true;
+          long curr = System.currentTimeMillis();
+          LOG.info(String.format(
+              "Finished fetch Namespace! %d secs used, numDirs = %d, numFiles = %d",
+              (curr - startTime) / 1000,
+              numDirectoriesFetched, numFilesFetched));
+        }
+        return;
+      }
+
+      try {
+        URIStatus status = fs.getStatus(new AlluxioURI(parent));
+        if (status != null && status.isFolder()) {
+          List<URIStatus> children = fs.listStatus(new AlluxioURI(parent));
+          FileStatusInternal internal = convertToFileStatusInternal(status, children.size());
+          internal.setPath(parent);
+          this.addFileStatus(internal);
+          numDirectoriesFetched++;
+
+          for (URIStatus child : children) {
+
+            if (child.isFolder()) {
+              this.deque.add(child.getPath());
+            } else {
+              this.addFileStatus(convertToFileStatusInternal(child, 0));
+              numFilesFetched++;
+            }
+          }
+        }
+      } catch (IOException | InterruptedException | AlluxioException e) {
+        LOG.error("Totally, numDirectoriesFetched = " + numDirectoriesFetched
+            + ", numFilesFetched = " + numFilesFetched
+            + ". Parent = " + parent, e);
+      }
+    }
+
+    public boolean finished() {
+      return this.isFinished;
+    }
+
+    public FileStatusInternalBatch pollBatch() {
+      return this.batches.poll();
+    }
+
+    public void addFileStatus(FileStatusInternal status) throws InterruptedException {
+      this.currentBatch.add(status);
+      if (this.currentBatch.isFull()) {
+        this.batches.put(currentBatch);
+        this.currentBatch = new FileStatusInternalBatch(DEFAULT_BATCH_SIZE);
+      }
+    }
+
+    private FileStatusInternal convertToFileStatusInternal(URIStatus status,
```

Comment:
Now `FileStatusInternal` is in `common` module and it's a subclass of `HdfsFileStatus`, which I think is not right now since alluxio also becomes a player. Maybe we should define a common pojo which has no dependency to any file system implementation.